intermediate_source/TP_tutorial.rst

```diff
@@ -0,0 +1,282 @@
+
+.. figure:: /_static/img/distributed/megatron_lm.png
+   :width: 100%
+   :align: center
+   :alt: Megatron-LM TP
+
+   Figure 1. represents the sharding in Tensor Parallelism style on a Transformer model’s MLP and Self-Attention layer, where the matmuls in both attention/MLP happens through sharded computations (`image source <https://arxiv.org/abs/1909.08053>__`)
+
+
+At a high level, PyTorch Tensor Parallel works as follows:
+
+*Sharding initialization*
+
+* Determine which ``ParallelStyle`` to apply to each layer and shard the initialized module by calling ``parallelize_module``
+* The parallelized modules would have their model parameters be swapped to DTensors, and DTensor would be responsible to run the parallelized module.
+
+*Runtime foward/backward*
+* Depending on the input/outputs DTensor layouts user specified for each parallel style, it would run proper communication operation to transform the DTensor layouts for inputs/outputs
+* Run sharded computation for the parallelized layers to save compute/memory. (i.e. ``nn.Linear``, ``nn.Embedding``)
+
+
+When and Why you should apply Tensor Parallel
+------------
+The PyTorch Fully Sharded Data Parallel (FSDP) could scale the model training to certain number of GPUs already, but when further scaling the model with regard to model size and number of GPUs, there are many more challenges need to be resolved, where it might need Tensor Parallelism (TP) to combine with FSDP:
+The world size (# GPUs) becoming too large (> 128/256 GPUs) and the FSDP collectives (i.e. ``allgather``) are being dominated by ring latency. TP/SP on top of FSDP would cut the FSDP world size by 8 thereby reducing the latency costs by 8.
+Hit data parallelism limit where you can not raise the global batch size to be above the # GPUs due to both convergence and GPU memory limitations, Tensor/Sequence Parallelism is the only known way to “ballpark” the global batch size and continue scaling with more GPUs. This means both model size and number of GPUs could continue to scale.
+For certain types of models, when local batch size becomes smaller TP/SP gives more flop friendly matrix multiplication shapes.
+
+So when pre-training, how easy is it to hit those limits? Today to pre-train a Large Language Model (LLM) with billions/trillions of tokens it could take months even with thousands of GPUs.
+It will always hit limitation 1 when training LLM on a large scale. For example, Llama 2 70B trained with 2k GPUs for 35 days, multi-dimensional parallelisms are needed at 2k scale.
```

Comment:
Here we are referring to **limitation 1** and **limitation 2**, shall we consider using numbered points for the limitations above?